src/category_theory/groupoid/subgroupoid.lean

```diff
@@ -0,0 +1,471 @@
+  assoc' := λ a b c d ⟨p,hp⟩ ⟨q,hq⟩ ⟨r,hr⟩, by simp only [category.assoc],
+  inv := λ a b p, ⟨inv p.val, S.inv' p.prop⟩,
+  inv_comp' := λ a b ⟨p,hp⟩, by simp only [inv_comp],
+  comp_inv' := λ a b ⟨p,hp⟩, by simp only [comp_inv] }
+
+/-- There is an embedding of the coerced subgroupoid to its parent-/
+def hom : (coe S) ⥤ C :=
+{ obj := λ c, c.val,
+  map := λ c d f, f.val,
+  map_id' := λ c, rfl,
+  map_comp' := λ c d e f g, rfl }
+
+lemma hom.inj_on_objects : function.injective (hom S).obj :=
+by { rintros ⟨c,hc⟩ ⟨d,hd⟩ hcd, simp only [subtype.mk_eq_mk], exact hcd }
+
+lemma hom.faithful :
+  ∀ c d, function.injective (λ (f : c ⟶ d), (hom S).map f) :=
+by { rintros ⟨c,hc⟩ ⟨d,hd⟩ ⟨f,hf⟩ ⟨g,hg⟩ hfg, simp only [subtype.mk_eq_mk], exact hfg, }
+
+/-- The subgroup of the vertex group at `c` given by the subgroupoid -/
+def vertex_subgroup {c : C} (hc : c ∈ S.objs) : subgroup (c ⟶ c) :=
+{ carrier  := S.arrows c c,
+  mul_mem' := λ f g hf hg, S.mul' hf hg,
+  one_mem' := id_mem_of_nonempty_isotropy _ _ hc,
+  inv_mem' := λ f hf, S.inv' hf }
+
+/-- A subgroupoid seen as a set of arrows -/
+def coe_set (S : subgroupoid C) : set (Σ (c d : C), c ⟶ d) := {F | F.2.2 ∈ S.arrows F.1 F.2.1 }
+
+private lemma mem_coe_set_iff' (S : subgroupoid C) {c d : C} (f : c ⟶ d) :
+  (⟨c,d,f⟩ : Σ (c d : C), c ⟶ d) ∈ S.coe_set ↔ f ∈ S.arrows c d := by refl
+
+instance : set_like (subgroupoid C) (Σ (c d : C), c ⟶ d) :=
+{ coe := coe_set,
+  coe_injective' := λ S T h, by
+  { ext c d f, simp_rw [←mem_coe_set_iff',h], } }
+
+@[simp] lemma mem_coe_set_iff (S : subgroupoid C) {c d : C} (f : c ⟶ d) :
+  (⟨c,d,f⟩ : Σ (c d : C), c ⟶ d) ∈ S ↔ f ∈ S.arrows c d := mem_coe_set_iff' S f
+
+@[simp] lemma le_iff (S T : subgroupoid C) : (S ≤ T) ↔ (∀ {c d}, (S.arrows c d) ⊆ (T.arrows c d)) :=
+begin
+  split,
+  { rintro h c d f, simp only [←mem_coe_set_iff], apply h, },
+  { rintro h ⟨c,d,f⟩, simp only [mem_coe_set_iff], apply h, },
+end
+
+instance : has_top (subgroupoid C) :=
+⟨ { arrows := (λ _ _, set.univ),
+    mul'   := by { rintros, trivial, },
+    inv'   := by { rintros, trivial, } } ⟩
+instance : has_bot (subgroupoid C) :=
+⟨ { arrows := (λ _ _, ∅),
+    mul'   := by { rintros, exfalso, assumption, },
+    inv'   := by { rintros, exfalso, assumption, } } ⟩
+
+instance : inhabited (subgroupoid C) := ⟨⊤⟩
+
+instance : has_inf (subgroupoid C) :=
+⟨ λ S T,
+  { arrows := (λ c d, (S.arrows c d) ∩ (T.arrows c d)),
+    inv'   := by { rintros, exact ⟨S.inv' hp.1, T.inv' hp.2⟩, },
+    mul'   := by { rintros, exact ⟨S.mul' hp.1 hq.1, T.mul' hp.2 hq.2⟩, } } ⟩
+
+instance : has_Inf (subgroupoid C) :=
+⟨ λ s,
+  { arrows := (λ c d, set.Inter (λ (S : s), S.val.arrows c d)),
+    inv'   := by
+    { rintros,
+      simp only [Inter_coe_set, mem_Inter] at hp ⊢,
+      exact λ S Ss, S.inv' (hp S Ss)},
+    mul'   := by
+    { rintros,
+      simp only [Inter_coe_set, mem_Inter] at hp hq ⊢,
+      exact λ S Ss, S.mul' (hp S Ss) (hq S Ss), } } ⟩
```

Comment:
```suggestion
  { arrows := λ c d, ⋂ S ∈ s, (subgroupoid.arrows S c d),
    inv' := by { intros, rw mem_Inter₂ at hp ⊢, exact λ S hS, S.inv' (hp S hS) },
    mul' := by { intros, rw mem_Inter₂ at hp hq ⊢,exact λ S hS, S.mul' (hp S hS) (hq S hS) } } ⟩
```
If Inf is used elsewhere you may have to fix things, but bounded binders like `⋂ S ∈ s` (which expand to double binders) are usually preferred over binders over a subtype.